### PR TITLE
 JWE Support

### PR DESCRIPTION
**Pull request by ortyomka**
Issue https://github.com/golang-jwt/jwe/issues/5

Add support of JWE with one key (Compact)

Add AES GCM cipher to encrypt content
Add RSA-OAEP to encrypt key

Add test from RFC7516 [Section 3.3](https://datatracker.ietf.org/doc/html/rfc7516#section-3.3) 

cc @mfridman @oxisto

### CODE REVIEW

**Comment by ortyomka** [comment]
@oxisto if you have free time, review the PR, please

**Comment by ortyomka** [comment]
I would like to finish this PR by Tuesday(12.07) inclusive.

If your schedule allows it, I'm willing to fix comments ASAP.

CC @mfridman @oxisto

**Comment by mfridman** [comment]
Hey there, thanks for putting a PR .. additions/changes related to security take time to review. It requires brushing up on the JWE specification, ensuring the implementation matches and there are no bugs or security issues to the best of our abilities/knowledge.

Furthermore, a lot of folks submit PR's, but then it's the maintainers who are on the hook to support that code and address bugs, issues, improvements, review new PR's, etc.

I can't speak for others, but right now is the end of quarter crunch time .. so my availability for open-source projects is limited. Tbh, adding JWE support isn't very high on my list and I'd rather not rush something for the sake of getting it out there.

**Comment by ortyomka** [comment]
I have understood your position. Thank you for your response.

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Copyright (c) 2012 Dave Grijalva
```
Comment (oxisto) [inline LICENSE]: I think we can just state (c) 2022 golang-jwt maintainers here

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,3 @@
+module github.com/golang-jwt/jwe
+
+go 1.16
```
Comment (oxisto) [inline go.mod]: Do you have any particular use case which requires Go 1.16? I was thinking whether, we could set this to 1.18 minimum so we could use generics / type constraints, possibly for the `key` value, as this is one of the biggest problems in the current JWT library.

File: jwe.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
+package jwe
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"strings"
+)
+
+// NewJWE creates a new JWE token.
+// The plaintext will be encrypted with the method using a cek(Content Encryption Key).
```
Comment (oxisto) [inline jwe.go]:
```suggestion
// The plaintext will be encrypted with the method using a Content Encryption Key (cek).
```

File: jwe.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	// Perform authenticated encryption on the plaintext
+	jwe.iv, jwe.ciphertext, jwe.tag, err = chipher.encrypt(cek, []byte(rawProtectedBase64), plaintext)
+	if err != nil {
+		return nil, err
+	}
+
+	return jwe, nil
+}
+
+type jwe struct {
```
Comment (oxisto) [inline jwe.go:55]: even though this is unexported, this is the most valuable struct and probably deserves a quick doc comment.

File: jwe_decrypt.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
+package jwe
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+)
+
+func (jwe jwe) Decrypt(key interface{}) ([]byte, error) {
```
Comment (oxisto) [inline jwe_decrypt.go:15]: Missing comment

File: jwe_decrypt.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+	method := jwe.protected.Enc
+	if len(method) == 0 {
+		return nil, errors.New("no \"enc\" header")
```
Comment (oxisto) [inline jwe_decrypt.go]: Can we extract the errors here in `ErrXX` values like the other files?

File: jwe_parse.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+func parseEncryptedCompact(input string) (*jwe, error) {
+	parts := strings.Split(input, ".")
+
+	if len(parts) != 5 {
+		return nil, errors.New("encrypted token contains an invalid number of segments")
+	}
+
+	jwe := &jwe{}
```
Comment (oxisto) [inline jwe_parse.go:27]: I guess we should avoid constructing this directly, if there is a `NewJWE` function, even though it is internal?

File: staticcheck.conf [hunk 8]
```diff
@@ -0,0 +1 @@
+checks = ["all", "-ST1000", "-ST1003", "-ST1016", "-ST1023"]
```
Comment (oxisto) [inline staticcheck.conf]: I think this is not necessary maybe c/p from the jwt repo

**Review by oxisto — COMMENTED** [review]
First set of review done. Thanks again for this contribution, it is very welcome. Sorry for the long delay in reviewing this.

There are still some kinks to work out in the (public) API, but since we are no way near releasing any 1.x version for this, it is ok and we can merge it once a few of the comments are addressed. Then afterwards, this probably needs another fine-tuning of the API.

File: jwe_parse.go (hunk 7 above)
Comment (ortyomka) [reply jwe_parse.go:27]: The `NewJWE` creates a new JWE with key and plaintext. Here is a JWE in compact form, we don't know key.
We work with final state and should parse it by parts.

File: staticcheck.conf (hunk 8 above)
Comment (ortyomka) [reply staticcheck.conf]: Removed

File: jwe_decrypt.go (hunk 6 above)
Comment (ortyomka) [reply jwe_decrypt.go]: Added

File: jwe_decrypt.go (hunk 5 above)
Comment (ortyomka) [reply jwe_decrypt.go:15]: Added

File: jwe.go (hunk 4 above)
Comment (ortyomka) [reply jwe.go:55]: Added

File: jwe.go (hunk 3 above)
Comment (ortyomka) [reply jwe.go]: Fixed

File: LICENSE (hunk 1 above)
Comment (ortyomka) [reply LICENSE]: Fixed

**Comment by ortyomka** [comment]
@oxisto Thank you for review. I fixed all comments

File: jwe_decrypt.go (hunk 5 above)
Comment (oxisto) [reply jwe_decrypt.go:15]: Small suggestion: ErrMissingEncHeader = errors.New(“missing \”enc\” header”)

**Comment by oxisto** [comment]
> @oxisto Thank you for review. I fixed all comments

Thank you! I have one small last comment for the errors and then I would suggest merging this initial version. Everything else can be done in increments.

File: jwe_decrypt.go (hunk 5 above)
Comment (ortyomka) [reply jwe_decrypt.go:15]: Fixed

**Comment by ortyomka** [comment]
@oxisto Added your suggestion

**Review by oxisto — APPROVED** [review]

**Pull request merged**